### ISSUE TITLE
Strange behavior when narrowing optional class members

### ISSUE THREAD

**Issue by matejcik** [microsoft/pyright#2679]
It seems that it's generally possible to narrow type of a variable through membership test.
The following snippet makes sure that `foo` is of type `str`, by checking that it is a member of `list[str]`.
```python
container: list[str] = []
foo: str | None

# this trickery is necessary to make foo actually be str | None
if __name__ == "__main__":
    foo = __name__ * 2
else:
    foo = None

reveal_type(foo)  # info: Type of "foo" is "str | None"

def takes_str(param: str) -> None:
    print(param)

if foo in container:
    reveal_type(foo)  # info: Type of "foo" is "str"
    takes_str(foo)
```

The same general idea (narrow optional type by checking membership in a non-optional collection) will fail if the variable is a class member:
```python
container: list[str] = []

def takes_str(param: str) -> None:
    print(param)

class A:
    def __init__(self) -> None:
        self.member: str | None = None

a = A()
reveal_type(a.member)  # info: Type of "a.member" is "str | None"

if a.member in container:
    reveal_type(a.member)  # info: Type of "a.member" is "str | None"
    takes_str(a.member)  # error: Argument of type "str | None" cannot be assigned to parameter "param" of type "str"
```

Curiously, adding an assert **after this check** will resolve the issue and the code type-checks:
```python
container: list[str] = []

def takes_str(param: str) -> None:
    print(param)


class A:
    def __init__(self) -> None:
        self.member: str | None = None

a = A()
reveal_type(a.member)    # info: Type of "a.member" is "str | None"

if a.member in container:
    reveal_type(a.member)  # info: Type of "a.member" is "str"
    takes_str(a.member)

if a.member in container:
    assert a.member is not None
```

**VS Code extension or command-line**
pyright **1.1.193**
`pyright --pythonversion=3.10 test.py`
I haven't tested other pythonversions.

**Comment by matejcik** [comment]
possible duplicate of #2666 ?

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.194, which I just published. It will also be included in the next release of pylance.
